utils/metrics/metrics.go

```diff
@@ -0,0 +1,187 @@
+		for categoryName, whiteListedValues := range config.Categories {
+			// Normalize to lower case since canonical for gRPC headers
+			// and convention for prometheus.
+			categoryName := strings.ToLower(categoryName)
+
+			// Store white listed category values as set for efficient access
+			whiteListedSet := make(map[string]struct{})
+			for _, categoryValue := range whiteListedValues {
+				whiteListedSet[categoryValue] = struct{}{}
+			}
+			categoryValues[categoryName] = whiteListedSet
+
+			// Construct a prometheus label for each category.
+			// Prometheus does not allow changing set of
+			// labels until next time bazel-remote is
+			// restarted.
+			labels = append(labels, categoryName)
+		}
+	}
+
+	// For now we only count AC requests, and only the most common status codes,
+	// becuse:
+	//
+	//  - No identified use case for others.
+	//  - Limit number of prometheus time series (if many configured categories).
+	//  - Reduce performance overhead of counters (if many configured categories).
+	//  - Would otherwise require injecting invocations in more places.
+	//
+	// But the naming, and the labels, of the counter, are generic to allow
+	// counting additional requests types or status codes in the future. Without
+	// having to rename the counter and get issues with non continous history of
+	// metrics.
+
+	counterIncomingCompletedReqs := promauto.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "bazel_remote_incoming_requests_completed_total",
+			Help: "The number of incoming HTTP and gRPC request. Currently only AC requests",
+		},
+		labels)
+
+	m := &metrics{
+		categoryValues:               categoryValues,
+		counterIncomingCompletedReqs: counterIncomingCompletedReqs,
+	}
+	return m
+}
+
+func getLabelValueFromHeaderValues(headerValues []string, whiteListedValues map[string]struct{}) string {
+	for _, headerValue := range headerValues {
+		// Prometheus only allows one value per label.
+		// Pick the first white listed header value we find.
+		if _, ok := whiteListedValues[headerValue]; ok {
+			return headerValue
+		}
+	}
+
+	// The values found in the header has not been white listed in
+	// the configuration file. Represent them as "other".
+	//
+	// The white listening is an attempt to avoid polluting
+	// prometheus with too many different time series.
+	//
+	// https://prometheus.io/docs/practices/naming/ warns about:
+	//
+	//   "CAUTION: Remember that every unique combination of key-value
+	//   label pairs represents a new time series, which can dramatically
+	//   increase the amount of data stored. Do not use labels to store
+	//   dimensions with high cardinality (many different label values),
+	//   such as user IDs, email addresses, or other unbounded sets of
+	//   values."
+	//
+	// It would have been nice if bazel-remote could reload the set
+	// of white listed values from updated configuration file, by
+	// SIGHUP signal instead of having to restart bazel-remote.
+	return "other"
```

Comment:
I wonder if this should be a word that gives more of a hint that it's masked a real user defined header value. I'm not sure of a good one, though

Reply:
Please tell if you find a good word!

I agree that ”other” can be confusing for someone trying to set a header.

But I hope the word ”other” would make sense for most people looking at a dashboard, without knowing that there is a concept of allowed values involved.

Reply:
Should this function return an error if the label isn't allowed?

Reply:
Do you mean reject the http request with an 4xx error? I think that would be too strict. See my long comment from today where I’m trying to elaborate on use cases.